lib/disposable.dart

```diff
@@ -179,16 +182,19 @@ class Disposable implements _Disposable, DisposableManager {
     }
     _isDisposing = true;
 
-    List<Future> futures = []
-      ..addAll(_internalDisposables.map((disposable) => disposable.dispose()))
+    // We need to filter out nulls because a subscription cancel
+    // method is allowed to return a plain old null value.
+    var futures = _blockingFutures
+      ..addAll(_internalDisposables
+          .map((disposable) => disposable.dispose())
+          .where((future) => future != null))
       ..add(onDispose());
 
     _internalDisposables = [];
+    _blockingFutures = new Set<Future>();
 
-    // We need to filter out nulls because a subscription cancel
-    // method is allowed to return a plain old null value.
     return Future
-        .wait(futures.where((future) => future != null))
+        .forEach(futures, (future) => future)
```

Comment:
They need to be ordered. I ended up disliking the `forEach` version because it was ugly. New version is much easier to understand.